resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala

```diff
@@ -180,6 +184,31 @@ private[spark] class Client(
       originalMetadata.setOwnerReferences(Collections.singletonList(driverPodOwnerReference))
     }
   }
+
+  // Build a Config Map that will house both the properties and the java options in a single file
+  private def buildConfigMap(
+    configMapName: String,
+    conf: SparkConf,
+    driverJavaOps: String): ConfigMap = {
+    val properties = new Properties()
+    conf
+      .remove(org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
+      .getAll.foreach { case (k, v) =>
+      properties.setProperty(k, v)
+    }
+    val propertiesWriter = new StringWriter()
+    properties.store(propertiesWriter,
+      s"Java properties built from Kubernetes config map with name: $configMapName")
+
+    val namespace = conf.get(KUBERNETES_NAMESPACE)
+    new ConfigMapBuilder()
+      .withNewMetadata()
+        .withName(configMapName)
+        .withNamespace(namespace)
+        .endMetadata()
+      .addToData(SPARK_CONF_FILE_NAME, propertiesWriter.toString + driverJavaOps)
```

Comment:
You want to write two config map data items:

- one data item with the properties file name,
- One data item with the JVM options.

Reply:
In other words I'd expect two calls to `addToData` here with the appropriate item names corresponding to the file names.

Reply:
We should probably also capture this with an integration test. Some test that has a call to `System.getProperty(property)` with a property that isn't a Spark-specific option should suffice.

Reply:
> One data item with the JVM options.

Why? spark-submit handles that already.

Reply:
This doesn't put `spark.driver.extraJavaOptions` on the launch of the spark-submit command in client mode, I think.

Reply:
Why does it have to be on the command line and not in the properties file?

Reply:
My understanding is for `spark-submit` to pick it up you want a `jvm-opts` file in the Spark configuration directory. This is specifically for JVM options that aren't `spark.*` options, e.g. GC, out of memory settings, yourkit agent loading.

Reply:
You really should try this out to see that it works the way I've been saying since the beginning. Just let spark-submit do its thing.

Reply:
Ah I think I was misreading the code here - just putting `spark.driver.extraJavaOptions` in the properties works. See https://github.com/apache/spark/blob/cfcd746689c2b84824745fa6d327ffb584c7a17d/launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java#L268

Sorry about that - @ifilonenko think just adding `spark.driver.extraJavaOptions` as-is to the properties file will suffice.